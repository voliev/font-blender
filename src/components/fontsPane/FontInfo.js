import React, { Fragment } from "react";
import classNames from "classnames";
import Preview from "@material-ui/icons/RemoveRedEyeRounded";
import PropTypes from "prop-types";

const FontInfo = ({
  family,
  category,
  subsets,
  variants,
  fetching,
  fontPreviewClickHandler
}) => {
  const fontFamilyStyle = {
    fontFamily: `'${family}', '${category}'`
  };

  const fontFamilyClasses = classNames({
    "family-name": true,
    fetching
  });

  return (
    <Fragment>
      <div className="family">
        <h4
          className={fontFamilyClasses}
          onClick={fontPreviewClickHandler}
          style={
            fetching
              ? {
                  fontFamily: `${category}`
                }
              : fontFamilyStyle
          }
          data-family={family}
          title="Preview font"
        >
          {family}
        </h4>
        <span className="preview-icon">
          <Preview fontSize="default" />
        </span>
      </div>
      <div className="category">{category}</div>
      <div className="charsets font_small">
        <span className="number">{subsets.length}</span> character{" "}
        {subsets.length === 1 ? "set" : "sets"}
      </div>
      <div className="styles font_small">
        <span className="number">{variants.length}</span>{" "}
        {variants.length === 1 ? "style" : "styles"}
      </div>
    </Fragment>
  );
};

FontInfo.propTypes = {
  family: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  variants: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  subsets: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  fetching: PropTypes.bool.isRequired,
  fontPreviewClickHandler: PropTypes.func.isRequired
};

export default FontInfo;
