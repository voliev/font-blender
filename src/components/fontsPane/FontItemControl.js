import React, { Component } from "react";
import classNames from "classnames";
import AddCircleRounded from "@material-ui/icons/AddCircleRounded";
import SyncRounded from "@material-ui/icons/SyncRounded";
import CheckCircleRounded from "@material-ui/icons/CheckCircleRounded";
import PropTypes from "prop-types";
import "./styles/FontItemControl.css";

class FontItemControl extends Component {
  getFontStateFromProps() {
    const { active, requested, selected, fetching } = this.props;
    const stateOptions = {
      selected,
      active: active && !selected,
      fetch: !active && requested && !fetching
    };

    for (const state in stateOptions) {
      if (stateOptions[state]) {
        return state;
      }
    }
  }

  getControlIconFromFontState(fontState) {
    const icons = {
      fetch: <SyncRounded fontSize="default" />,
      active: <AddCircleRounded fontSize="default" />,
      selected: <CheckCircleRounded fontSize="default" />
    };

    return icons[fontState];
  }

  getTitleAttributeFromFontState(fontState) {
    const titleAttributes = {
      fetch: "Font has not been loaded. Try again",
      active: "Add to Selected",
      selected: "Deselect font"
    };

    const title = titleAttributes[fontState] || "Fetching...";

    return title;
  }

  render() {
    const {
      family,
      requested,
      prefetched,
      fetched,
      fetching,
      selected,
      fontSelectionDeselectionHandler,
      requestFont
    } = this.props;

    const componentClasses = classNames({
      FontItemControl: true,
      fetching
    });
    const btnClasses = classNames({
      btn: true,
      "btn-round-small": true,
      "add-circle": !selected && (prefetched || fetched),
      checked: selected && !fetching,
      sync: requested && !prefetched && !fetched
    });

    const fontState = this.getFontStateFromProps();

    return (
      <div className={componentClasses}>
        <button
          className={btnClasses}
          onClick={prefetched ? fontSelectionDeselectionHandler : requestFont}
          data-family={family}
          value={family}
          title={this.getTitleAttributeFromFontState(fontState)}
        >
          {this.getControlIconFromFontState(fontState)}
        </button>
      </div>
    );
  }
}

FontItemControl.propTypes = {
  family: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
  requested: PropTypes.bool.isRequired,
  prefetched: PropTypes.bool.isRequired,
  fetched: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  selected: PropTypes.bool.isRequired,
  fontSelectionDeselectionHandler: PropTypes.func.isRequired,
  requestFont: PropTypes.func.isRequired
};

export default FontItemControl;
