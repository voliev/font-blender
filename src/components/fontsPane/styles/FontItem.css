.FontItem {
  display: grid;
  grid-template-columns: 8rem 8rem 2rem;
  grid-template-rows: minmax(2rem, max-content) 1.5rem 1.5rem;
  grid-template-areas:
    "family family controls"
    "category category controls"
    "charsets styles controls";
  cursor: default;
  opacity: 1;
  -webkit-transition: opacity 0.8s, background-color 0.2s;
  transition: opacity 0.8s, background-color 0.2s;
  padding: var(--space-md) var(--space-md) var(--space-md) var(--space-lg);

  position: relative;
}

.FontItem.selected,
.FontItem:hover {
  transition: background-color 0.2s, color 0.2s;
  background-color: var(--almost-white);
}

.FontItem.selected::after,
.FontItem:hover::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  width: .4rem;
  height: 100%;
  background-color: var(--blue);
  animation: growY .1s ease-out 1 forwards;

  transform-origin: 50% 50%;
}

.FontItem .family {
  align-items: flex-start;
  cursor: pointer;
  display: flex;
  grid-area: family;
  justify-content: space-between;
  margin-top: 0;
}

.FontItem .family .family-name {
  /* TODO: Evaluate margin benefits in this case */
  margin: 0 var(--space-xs) 0 0;
}

.FontItem .family .preview-icon {
  align-items: center;
  align-self: flex-start;
  color: var(--grey);
  display: flex;
  opacity: 0;
  transition: opacity 0.1s;
}

.FontItem.preview .family .preview-icon,
.FontItem .family-name:hover~.preview-icon {
  opacity: 1;
  transition: opacity 0.1s;
}

.FontItem.preview .family .preview-icon {
  color: var(--blue);
}

.FontItem .category {
  align-self: center;
  color: var(--dark-grey);
  grid-area: category;
  justify-self: left;
  font-weight: 700;
}

.FontItem .charsets {
  align-self: center;
  color: var(--dark-grey);
  grid-area: charsets;
  justify-self: left;
}

.FontItem .styles {
  align-self: center;
  color: var(--dark-grey);
  grid-area: styles;
  justify-self: left;
}

.FontItem .number {
  color: var(--dark-grey);
  font-weight: 700;
}

.FontItem.empty {
  opacity: 0.4;
  transition: opacity 0.8s;
}

.FontItem.fetching>*,
.FontItem.blank>* {
  filter: blur(6px);
}
