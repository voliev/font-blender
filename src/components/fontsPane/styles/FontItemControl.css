.FontItemControl {
  align-self: stretch;
  grid-area: controls;
  justify-self: end;
}

.FontItemControl.fetching {
  color: transparent;
  text-shadow: var(--blur-text);
}

.FontItemControl .btn {
  color: var(--light-grey);
  margin: 0;
  transition: color 0.1s ease-out, transform 0.1s ease-out, box-shadow 0.1s;
}

.FontItemControl .btn:hover {
  transform: scale(1.2)
}

.FontItemControl .btn.add-circle:hover {
  color: var(--green);
}

.FontItemControl .btn:focus {
  outline: none;
}

.FontItemControl .btn.add-circle,
.FontItemControl .btn.sync {
  color: var(--grey);
}

.FontItemControl .btn.checked {
  color: var(--green);
}

.FontItemControl:not(.fetching) .btn.sync {
  color: var(--orange);
}

.FontItemControl:not(.fetching) .btn.sync:hover {
  transform: scale(1.2);
  transform-origin: .75rem .75rem;
}

.FontItemControl:not(.fetching) .btn.sync:active {
  transform: rotate(-.25turn);
  transform-origin: .75rem .75rem;
}
