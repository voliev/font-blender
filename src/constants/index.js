// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
