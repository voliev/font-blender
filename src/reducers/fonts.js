import { combineReducers } from "redux";
import {
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fontsDataStatus = (state = "", action) => {
  switch (action.type) {
    case FETCH_FONTS_DATA:
      return "fetching";
    case FETCH_FONTS_DATA_SUCCESS:
      return "success";
    case FETCH_FONTS_DATA_FAILURE:
      return "failure";
    default:
      return state;
  }
};

const initialState = {};

export const all = (state = initialState, action) => {
  switch (action.type) {
    case SET_FONTS:
      return action.fonts;
    case UPDATE_FONT:
      return {
        ...state,
        [action.family]: {
          ...state[action.family],
          ...action.updates
        }
      };
    default:
      return state;
  }
};

export default combineReducers({
  all,
  fontsDataStatus
});

export const getFonts = state => state.fonts.all;
export const getFont = (state, family) => state.fonts.all[family];
export const getFontsDataStatus = state => state.fonts.fontsDataStatus;
