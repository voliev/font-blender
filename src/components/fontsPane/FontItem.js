import React, { PureComponent } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import { CHAR_RANGE } from "../../constants";
import FontInfo from "./FontInfo";
import FontItemControl from "./FontItemControl";
import "./styles/FontItem.css";

class FontItem extends PureComponent {
  constructor() {
    super();

    this.fontItem = React.createRef();

    this.state = {
      fontItem: undefined
    };

    this.requestFont = this.requestFont.bind(this);
    this.handleFontSelectionDeselection = this.handleFontSelectionDeselection.bind(
      this
    );
    this.handleFontPreviewClick = this.handleFontPreviewClick.bind(this);
  }

  requestFont() {
    this.props.fetchFont(this.props.font.family, CHAR_RANGE);
  }

  handleFontSelectionDeselection(e) {
    const { value } = e.currentTarget;
    !this.props.selected
      ? this.props.addFontToSelected(value)
      : this.props.removeFontFromSelected(value);
  }

  handleFontPreviewClick(e) {
    const { family } = e.currentTarget.dataset;
    this.props.startDisplayShowcaseFontPreview(family);
  }

  componentDidUpdate() {
    if (this.props.isVisible && !this.props.font.requested) {
      this.requestFont();
    }
  }

  componentDidMount() {
    this.setState({ fontItem: this.fontItem.current });
  }

  render() {
    const { font, selected, previewFontFamily } = this.props;

    const componentClasses = classNames({
      FontItem: true,
      selected,
      fetching: font.fetching,
      blank: !font.requested,
      preview: font.family === previewFontFamily
    });

    return (
      <li className={componentClasses} ref={this.fontItem}>
        <FontInfo
          {...font}
          fontPreviewClickHandler={this.handleFontPreviewClick}
        />
        <FontItemControl
          {...font}
          selected={selected}
          fontSelectionDeselectionHandler={this.handleFontSelectionDeselection}
          requestFont={this.requestFont}
        />
      </li>
    );
  }
}

FontItem.propTypes = {
  font: PropTypes.exact({
    family: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    variants: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    subsets: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    fetching: PropTypes.bool.isRequired,
    active: PropTypes.bool.isRequired,
    requested: PropTypes.bool.isRequired,
    prefetched: PropTypes.bool.isRequired,
    fetched: PropTypes.bool.isRequired
  }).isRequired,
  selected: PropTypes.bool.isRequired,
  isVisible: PropTypes.bool,
  previewFontFamily: PropTypes.string,
  fetchFont: PropTypes.func.isRequired,
  addFontToSelected: PropTypes.func.isRequired,
  removeFontFromSelected: PropTypes.func.isRequired,
  startDisplayShowcaseFontPreview: PropTypes.func.isRequired
};

export default FontItem;
