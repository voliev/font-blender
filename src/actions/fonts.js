import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
