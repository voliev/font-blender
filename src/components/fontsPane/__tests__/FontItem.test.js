import React from "react";
import { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import FontItem from "../FontItem";
import FontInfo from "../FontInfo";
import FontItemControl from "../FontItemControl";
import { CHAR_RANGE } from "../../../constants";
import fonts from "../../../fixtures/fonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    font: fonts["Roboto"],
    isVisible: false,
    selected: false,
    previewFontFamily: undefined,
    fetchFont: jest.fn(),
    addFontToSelected: jest.fn(),
    removeFontFromSelected: jest.fn(),
    startDisplayShowcaseFontPreview: jest.fn(),
    ...propsOverrides
  };

  const mockStore = configureStore();
  const store = mockStore();

  return {
    ...props,
    wrapper: renderFn(<FontItem {...props} store={store} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const family = "Open Sans";
  const font = fonts[family];
  const { wrapper } = setup(shallow, { font });

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const family = "Open Sans";
  const font = fonts[family];
  const {
    wrapper,
    fetchFont,
    addFontToSelected,
    removeFontFromSelected,
    startDisplayShowcaseFontPreview
  } = setup(mount, { font });

  expect(typeof wrapper.prop("font")).toBe("object");
  expect(wrapper.prop("font")).toEqual(font);

  expect(typeof wrapper.prop("isVisible")).toBe("boolean");
  expect(wrapper.prop("isVisible")).toBeFalsy();

  expect(typeof wrapper.prop("selected")).toBe("boolean");
  expect(wrapper.prop("selected")).toBeFalsy();

  expect(wrapper.prop("previewFontFamily")).toBeUndefined();

  expect(typeof wrapper.prop("fetchFont")).toBe("function");
  expect(wrapper.prop("fetchFont")).toBe(fetchFont);

  expect(typeof wrapper.prop("addFontToSelected")).toBe("function");
  expect(wrapper.prop("addFontToSelected")).toBe(addFontToSelected);

  expect(typeof wrapper.prop("removeFontFromSelected")).toBe("function");
  expect(wrapper.prop("removeFontFromSelected")).toBe(removeFontFromSelected);

  expect(typeof wrapper.prop("startDisplayShowcaseFontPreview")).toBe(
    "function"
  );
  expect(wrapper.prop("startDisplayShowcaseFontPreview")).toBe(
    startDisplayShowcaseFontPreview
  );
});

it("should request font when list item becomes visible", () => {
  const family = "Roboto";
  const font = fonts[family];
  const { wrapper, fetchFont } = setup(mount, { font });

  expect(fetchFont).not.toHaveBeenCalled();

  wrapper.setProps({ isVisible: true });

  expect(fetchFont).toHaveBeenCalledTimes(1);
  expect(fetchFont).toHaveBeenCalledWith(family, CHAR_RANGE);
});

it("should set corresponding component classes", () => {
  const family = "Roboto";
  const font = fonts[family];
  const { wrapper } = setup(shallow, { font });

  expect(wrapper.hasClass("FontItem")).toBeTruthy();
  expect(wrapper.hasClass("selected")).toBeFalsy();
  expect(wrapper.hasClass("fetching")).toBeFalsy();
  expect(wrapper.hasClass("blank")).toBeTruthy();
  expect(wrapper.hasClass("preview")).toBeFalsy();

  wrapper.setProps({
    font: {
      ...font,
      fetching: true
    }
  });
  expect(wrapper.hasClass("fetching")).toBeTruthy();

  wrapper.setProps({
    font: {
      ...font,
      requested: true
    }
  });
  expect(wrapper.hasClass("blank")).toBeFalsy();

  wrapper.setProps({ selected: true });

  expect(wrapper.hasClass("selected")).toBeTruthy();

  wrapper.setProps({ previewFontFamily: family });

  expect(wrapper.hasClass("preview")).toBeTruthy();
});

it("should handle font preview click", () => {
  const family = "Open Sans";
  const font = fonts[family];
  const { wrapper, startDisplayShowcaseFontPreview } = setup(mount, { font });

  const event = { currentTarget: { dataset: { family } } };
  wrapper.find(FontInfo).prop("fontPreviewClickHandler")(event);

  expect(startDisplayShowcaseFontPreview).toHaveBeenCalledTimes(1);
  expect(startDisplayShowcaseFontPreview).toHaveBeenCalledWith(family);
});

it("should handle font selection/deselection", () => {
  const family = "Open Sans";
  const font = fonts[family];
  const { addFontToSelected, removeFontFromSelected, wrapper } = setup(
    shallow,
    { font }
  );

  const event = { currentTarget: { value: family } };
  wrapper.find(FontItemControl).prop("fontSelectionDeselectionHandler")(event);

  expect(addFontToSelected).toHaveBeenCalledTimes(1);
  expect(addFontToSelected).toHaveBeenCalledWith(family);
  expect(removeFontFromSelected).toHaveBeenCalledTimes(0);

  wrapper.setProps({ selected: true });
  wrapper.find(FontItemControl).prop("fontSelectionDeselectionHandler")(event);

  expect(removeFontFromSelected).toHaveBeenCalledTimes(1);
  expect(removeFontFromSelected).toHaveBeenCalledWith(family);
  expect(addFontToSelected).toHaveBeenCalledTimes(1);
});
