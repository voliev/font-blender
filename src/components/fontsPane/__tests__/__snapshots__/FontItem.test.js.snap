// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<li
  className="FontItem"
>
  <FontInfo
    active={true}
    category="sans-serif"
    family="Open Sans"
    fetched={true}
    fetching={false}
    fontPreviewClickHandler={[Function]}
    prefetched={true}
    requested={true}
    subsets={
      Array [
        "greek",
        "vietnamese",
        "latin-ext",
        "cyrillic-ext",
        "greek-ext",
        "latin",
        "cyrillic",
      ]
    }
    variants={
      Array [
        "300",
        "300italic",
        "regular",
        "italic",
        "600",
        "600italic",
        "700",
        "700italic",
        "800",
        "800italic",
      ]
    }
  />
  <FontItemControl
    active={true}
    category="sans-serif"
    family="Open Sans"
    fetched={true}
    fetching={false}
    fontSelectionDeselectionHandler={[Function]}
    prefetched={true}
    requestFont={[Function]}
    requested={true}
    selected={false}
    subsets={
      Array [
        "greek",
        "vietnamese",
        "latin-ext",
        "cyrillic-ext",
        "greek-ext",
        "latin",
        "cyrillic",
      ]
    }
    variants={
      Array [
        "300",
        "300italic",
        "regular",
        "italic",
        "600",
        "600italic",
        "700",
        "700italic",
        "800",
        "800italic",
      ]
    }
  />
</li>
`;
